### PR TITLE
contract auth

### PR DESCRIPTION
**Pull request by tbfleming**
Add contract auth support

Phase 1 (this PR): Only actions which support session keys are dispatchable through the `run` action. All actions support the normal dispatch.

Phase 2 (future): All actions will be dispatchable through the `run` action. Ones which don't support session keys will require `account_auth` or `no_auth`; they won't support `signature_auth`. TBD if this really happens.

Phase 3 (future): Normal dispatch will be disabled for actions other than `newsession` and `run`. TBD if this really happens.

### CODE REVIEW

File: libraries/eosiolib/contracts/include/eosio/action.hpp [hunk 1]
```diff
@@ -54,6 +54,28 @@ namespace eosio
       }
    };  // namespace internal_use_do_not_use
 
+   template <typename T>
+   struct excluded_arg
```
Comment (swatanabe) [inline libraries/eosiolib/contracts/include/eosio/action.hpp:58]: The name `excluded_arg` is confusing especially when combined with `ignore` which sounds vaguely similar but means almost the opposite.

File: contracts/eden/include/eden.hpp [hunk 2]
```diff
@@ -74,6 +79,20 @@ namespace eden
...
+                      const std::string& description);
+
+      void delsession(const eosio::excluded_arg<session_info>& current_session,
+                      eosio::name eden_account,
+                      const eosio::public_key& key);
+
+      void execsession(const eosio::signature& signature,
+                       eosio::ignore<eosio::name> eden_account,
+                       eosio::ignore<eosio::varuint32> sequence,
+                       eosio::ignore<std::vector<action>> actions);
```
[the remark this answers is not in the document]
Comment (tbfleming) [reply contracts/eden/include/eden.hpp:94]: If we come up with an alternative, I don't want it to sacrifice the ability of existing history tools to create a human-readable decoding of the original request. Right now they can use ABI decoding to show what's in the action vector.

[the remark this answers is not in the document]
Comment (swatanabe) [reply contracts/eden/include/eden.hpp:94]: For history purposes,I don't see why it's important that the original request is human readable vs. having a human readable decoding in the trace.  Having the original request human readable is normally critical for signing, but that's not an issue here.

File: contracts/eden/src/actions/sessions.cpp [hunk 3]
```diff
@@ -157,7 +166,7 @@ namespace eden
          auto& sequences = session->sequences;
          if (sequences.begin() != sequences.end())
          {
-            if (sequence.value < *sequences.begin())
+            if (sequence.value < *sequences.begin() && sequences.size() < 20)
```
Comment (swatanabe) [inline contracts/eden/src/actions/sessions.cpp:169]: The extra condition is backwards.

File: contracts/eden/include/eden.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (sparkplug0025) [reply contracts/eden/include/eden.hpp:94]: > a human readable decoding in the trace
starting to look good... but still hard to digest in block explorer (needs to open the transactions first and then look for traces)

File: contracts/eden/src/actions/sessions.cpp [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+namespace eden
+{
+   void expire(eosio::name contract, session_container& sc)
+   {
+      auto now = eosio::current_block_time();
+      auto& sessions = sc.sessions();
+      for (auto& session : sessions)
+         if (session.expiration <= now)
+            push_event(session_del_event{sc.eden_account(), session.key}, contract);
```
[the remark this answers is not in the document]
Comment (tbfleming) [reply contracts/eden/src/actions/sessions.cpp:22]: I'm split. It makes sense for the history side to be aware of that, but we were bitten recently by the history side knowing those things.

**Review by swatanabe — APPROVED** [review]

**Pull request merged**